pkg/app/piped/eventwatcher/eventwatcher.go

```diff
@@ -0,0 +1,254 @@
+				w.logger.Error("failed to update the values",
+					zap.String("repo-id", repoCfg.RepoID),
+					zap.Error(err),
+				)
+			}
+		}
+	}
+}
+
+// updateValues inspects all Event-definition and pushes the changes to git repo if there is.
+func (w *watcher) updateValues(ctx context.Context, repo git.Repo, events []config.EventWatcherEvent, commitMsg string) error {
+	commits := make([]*commit, 0)
+	for _, e := range events {
+		c, err := w.checkOutdatedValues(ctx, &e, repo, commitMsg)
+		if err != nil {
+			w.logger.Error("failed to check outdated value", zap.Error(err))
+			continue
+		}
+		if c != nil {
+			commits = append(commits, c)
+		}
+	}
+	if len(commits) == 0 {
+		return nil
+	}
+
+	w.logger.Info(fmt.Sprintf("there are %d outdated values", len(commits)))
+	// Copy the repo to another directory to avoid pull failure in the future.
+	tmpDir, err := ioutil.TempDir("", "event-watcher")
+	if err != nil {
+		return fmt.Errorf("failed to create a new temporary directory: %w", err)
+	}
+	defer os.RemoveAll(tmpDir)
+	tmpRepo, err := repo.Copy(filepath.Join(tmpDir, "tmp-repo"))
+	if err != nil {
+		return fmt.Errorf("failed to copy the repository to the temporary directory: %w", err)
+	}
+	for _, c := range commits {
+		if err := tmpRepo.CommitChanges(ctx, tmpRepo.GetClonedBranch(), c.message, false, c.changes); err != nil {
+			return fmt.Errorf("failed to perform git commit: %w", err)
+		}
+	}
+
+	return tmpRepo.Push(ctx, tmpRepo.GetClonedBranch())
+}
+
+// checkOutdatedValues gives back a change contents if any deviation exists between the value in
+// the given git repository and one in the control-plane.
+func (w *watcher) checkOutdatedValues(ctx context.Context, event *config.EventWatcherEvent, repo git.Repo, commitMsg string) (*commit, error) {
+	latestEvent, ok := w.eventGetter.GetLatest(ctx, event.Name, event.Labels)
+	if !ok {
+		return nil, fmt.Errorf("failed to get the latest Event with the name %q", event.Name)
+	}
+
+	// Determine files to be changed.
+	changes := make(map[string][]byte, 0)
+	for _, r := range event.Replacements {
+		path := filepath.Join(repo.GetPath(), r.File)
+		yml, err := ioutil.ReadFile(path)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read file: %w", err)
+		}
+		v, err := yamlprocessor.GetValue(yml, r.YAMLField)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get value at %s in %s: %w", r.YAMLField, r.File, err)
+		}
+		value, ok := v.(string)
+		if !ok {
+			return nil, fmt.Errorf("unknown type of value is defined at %s in %s", r.YAMLField, r.File)
+		}
+		if latestEvent.Data == value {
+			// Already up-to-date.
+			continue
+		}
+		// Edit the local file and put it into the change list.
+		newYml, err := yamlprocessor.ReplaceValue(yml, r.YAMLField, latestEvent.Data)
+		if err != nil {
+			return nil, fmt.Errorf("failed to replace value at %s with %s: %w", r.YAMLField, latestEvent.Data, err)
+		}
+		if err := ioutil.WriteFile(path, newYml, os.ModePerm); err != nil {
```

Comment:
I have just realized that we have to save it in memory instead of writing to the local path.
Because writing it back to the local path will cause an error while doing the next pulls.

Reply:
So checking it in the `changes` map before loading it from the local path looks good.

Reply:
You are right! using cache looks good. Alternatively, I think we can leverage the temp repo because we use it when pushing. We can use it when making changes as well
https://github.com/pipe-cd/pipe/pull/1411/files#diff-60908ffb7e37bfc3692d8c4cace4be91d53bccf9a549fb91ea30b70387abd9cfR191

Reply:
(Totally I misunderstood that we're using tmpRepo when making changes.)

Reply:
Hmm... using tmp repo makes the logic simple. But on the dark side, copying the directory happens every time no matter commits exits. After thinking about that, I'm begging to think the way to check `changes` you suggested is more appropriate.

Reply:
Checking `changes` makes it a little bit tougher to read, but looks not bad.
```go
		var err error
		yml, ok := changes[r.File]
		if !ok {
			path := filepath.Join(repo.GetPath(), r.File)
			yml, err = ioutil.ReadFile(path)
			if err != nil {
				return nil, fmt.Errorf("failed to read file: %w", err)
			}
		}
		v, err := yamlprocessor.GetValue(yml, r.YAMLField)
```

It's completely ok, but IMO we should act as naturally as possible. Using tmpRepo doesn't look that bad. It would be great if you give me any opinion.

Reply:
IMO, checking `changes` looks better.

Reply:
Good. Let me do so.

Reply:
Hold on, I found it is still likely to overwritten when processing other events. The `changes` is reset for each event. So looks like it's safer to use tmpRepo.

https://github.com/pipe-cd/pipe/pull/1411/files#diff-60908ffb7e37bfc3692d8c4cace4be91d53bccf9a549fb91ea30b70387abd9cfR170-R171

Reply:
Right. lol. So let's update the local data. And please change the function name to match its new behavior.

Reply:
Got it. The current one is absolutely not appropriate.